### DISCUSSION TITLE
Optional scopes not getting included: React SPA (MSAL) + Keycloak + Django Backend

### DISCUSSION THREAD

**Discussion by sachingaikwad123** [keycloak/keycloak#33133] [Q&A]
Hello everyone,

I have this typical setup where I have:
Frontend: React SPA (using MSAL)
IAM: Keycloak
Backend: Django DRF (REST APIs and GraphQL APIs)

TLDR: I am requesting "scope" (for backend api access) from frontend (SPA), but not getting that "scope" in "Access Token" received.

### Requirement:

1. I would like to authenticate users in 'Frontend' using Keycloak SSO. (Current status: Works fine with PKCE with Authorization Code Flow for Public Client)
2. After 'login' authentication is successful, 'Frontend' can not ask for 'Access Token' for different 'scopes' (different applications/clients which are registered with Keycloak) so that it can call their Web-APIs (REST APIs or GraphQL APIs). In this case, Django Backend is another client registered with Keycloak. So, 'Frontend' should get 'access token' with 'scope'. [Current Status: I have a problem! ]
3. Use this 'Access Token' in axios/react-query clients 'Header' to access REST APIs and GraphQL APIs in Django Backend. [Current Status: Not reached here yet]

Note: I had got my setup working with AAD (Microsoft Entra ID) (free license). So, I know that my code works in one setup. There must be something keycloak specific configuration which I am missing.

### What is happening?

1. React SPA (Frontend) does login with 'msal.loginRedirect()' and default scopes ["openid", "profile", "email"]. This succeeds and React SPA gets 'ID Token' and 'Access Token'. Note: This access token is for different scope.
2. Now, React SPA (Frontend) calls 'msal.acquireTokenSilent()' with scope: ["restapi"]. This also succeeds and we get 'Access Token'. But this 'Access Token' does not have "restapi" in "scope" parameter. [Problem!]

### Setup Details

1. User: test@myrealm.in
2. Group: mygroup

- Role Mapping: 
- -     frontend-client: Admin
- -     backend-client: Admin

3. Client Scopes

- restapi
- - Scope: [I am not sure if I need to assign any 'Role' here. I played around with it. But did not get satisfactory result.]
- - - Assigning 'backend-client: Admin' is preferred here. Only after adding this, I get 'resource_access.backend-client.role' in scope when I try things from "Evaluate" in Keycloak UI.

4. frontend-client:

- Roles: Admin
- Client Scopes: Added 'restapi' as 'Optional' scope
- [frontend-client-dedicated]: Added 'client-role-mapper' to include 'resource_access.${client_id}.roles'. 'Full scope Allowed' is turned OFF.

5. backend-client:

- Roles: Admin
- Client Scopes: Added 'restapi' as 'Optional' scope
- [backend-client-dedicated]: Added 'client-role-mapper' to include 'resource_access.${client_id}.roles'. 'Full scope Allowed' is turned OFF.

### Observations

1. Keycloak UI: When I try with "Evaluate" option and ask for "restapi" scope, I see that "Access Token" as "restapi" scope for "test" user.
2. For "test" user, I see that "Consents" tab has this: "Granted client scopes" is empty. I think this shows that there is a problem.

What am I missing in my setup? Please suggest.

**Comment by sachingaikwad123** [comment]
Update: When I change 'restapi' [Client Scope] to "Default" it is returned for "test" user.

**Comment by sachingaikwad123** [comment]
Here is what I think is happening.

1. When I am calling MSAL.acquireTokenSilent() it is sending token request (POST) with 'grant_type = refresh_token'. Meaning, it is trying to get new 'access token' by providing already issued 'refresh token'.
2. In 'refreshAccessToken' implementation of keycloak, it filters out any new scope which may have been asked by 'client'. It makes sure that it keeps only scope which are present in 'refresh token' (initial scopes). Comment from code: **"The requested scope MUST NOT include any scope not originally granted by the resource owner"**. I think this is why I don't see 'restapi' scope in my 'access token'.

```
    public AccessTokenResponseBuilder refreshAccessToken(KeycloakSession session, UriInfo uriInfo, ClientConnection connection, RealmModel realm, ClientModel authorizedClient,
                                            String encodedRefreshToken, EventBuilder event, HttpHeaders headers, HttpRequest request, String scopeParameter) throws OAuthErrorException {
        RefreshToken refreshToken = verifyRefreshToken(session, realm, authorizedClient, request, encodedRefreshToken, true);

        event.session(refreshToken.getSessionState())
                .detail(Details.REFRESH_TOKEN_ID, refreshToken.getId())
                .detail(Details.REFRESH_TOKEN_TYPE, refreshToken.getType());

        if (refreshToken.getSubject() != null) {
            event.detail(Details.REFRESH_TOKEN_SUB, refreshToken.getSubject());
        }

        // Setup clientScopes from refresh token to the context
        String oldTokenScope = refreshToken.getScope();
        //The requested scope MUST NOT include any scope not originally granted by the resource owner
        //if scope parameter is not null, remove every scope that is not part of scope parameter
        if (scopeParameter != null && ! scopeParameter.isEmpty()) {
            Set<String> scopeParamScopes = Arrays.stream(scopeParameter.split(" ")).collect(Collectors.toSet());
            oldTokenScope = Arrays.stream(oldTokenScope.split(" "))
                    .map(transformScopes(scopeParamScopes))
                    .filter(Objects::nonNull)
                    .collect(Collectors.joining(" "));
        }
```

**Questions**

1. Can anyone from keycloak team confirm if my understanding is right about this case?
2. Why does this work in Azure AAD? Why acquireTokenSilent() is able to exchange 'initial refresh token' and get new 'access token' with a different scope there? Why Azure AAD decided to allow this? or is there any configuration in clients in registration which "allows" this?
3. It seems like 'Keycloak' is doing the right that. Clients should not be allowed to get new 'access token' with different scope if initial 'refresh token' is from different scope. But, what is the way for clients to get 'access token' with any new 'scope'? Is 'Token Exchange' only way for this?

**Reply by sachingaikwad123** [reply]
@cgeorgilakis and @mposolda Could you please take a look at this?

I understand that Keycloak implementation is right and as per OAuth RFC. But, I am wondering how to request 'access token' for 'new scope for new resource' in that case?

P.S.: I am using MSAL library in my React SPA.

**Comment by cgeorgilakis** [comment]
According to [Section 6 in OAuth2 RFC](https://www.ietf.org/rfc/rfc6749.txt):

> scope
> OPTIONAL. The scope of the access request as described by
> Section 3.3. The requested scope MUST NOT include any scope
> not originally granted by the resource owner, and if omitted is
> treated as equal to the scope originally granted by the
> resource owner.

So, Keycloak behaviour is based on OAuth2 protocol and it is the logical behaviour.
Why not adding restapi scope in the first call?

**Reply by sachingaikwad123** [reply]
Yes. I agree that Keycloak's behavior is correct and as per RFC.

> Why not adding restapi scope in the first call?

1. In the first-call, I want to get only IDToken which is valid for 'frontend-client'. So, 'aud' should be 'frontend-client' only. If I include 'restapi' scope in this, then I have 'audience-backend-mapper' for 'restapi' scope which writes 'aud = backend-client' in the token claims. After SSO login, I would like 'IDToken' (and AccessToken) to have 'aud = frontend-client'. Because that's what these token are meant for.
2. Why am I having this 'audience-backend-mapper' in 'restapi' scope? I am getting this 'Access Token' for 'backend api'. When I will send this 'Access Token' to backend (Django), it will validate that will expect 'aud' to have 'backend-client' in it - i.e. this toke is for me. Otherwise, this validation will fail, right?

Please let me know if I should configure things differently.

**Reply by sachingaikwad123** [reply]
In other words, there must be a way for 'React SPA' (frontend client) to get multiple 'access token' for different 'optional scopes' from whenever required (on demand), right? How to do this?
